### ISSUE TITLE
2.1.20365: UnitSeparator#categorize:103 - ConcurrentModificationException

### ISSUE THREAD

**Issue by tripleabuilderbot** [triplea-game/triplea#7066]
## Map
A Son of Ice and Fire

## TripleA Version
2.1.20365

## Java Version
11.0.6

## Operating System
Windows 10

## Stack Trace
```
java.util.ConcurrentModificationException
	at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1042)
	at java.base/java.util.ArrayList$Itr.next(ArrayList.java:996)
	at java.base/java.util.Collections$UnmodifiableCollection$1.next(Collections.java:1047)
	at games.strategy.triplea.util.UnitSeparator.categorize(UnitSeparator.java:103)
	at games.strategy.triplea.util.UnitSeparator.categorize(UnitSeparator.java:68)
	at games.strategy.triplea.util.UnitSeparator.categorize(UnitSeparator.java:153)
	at games.strategy.triplea.util.UnitSeparator.categorize(UnitSeparator.java:59)
	at games.strategy.triplea.ui.PlacementUnitsCollapsiblePanel.lambda$updateStep$0(PlacementUnitsCollapsiblePanel.java:42)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

```

**Comment by asvitkine** [comment]
Looks like the problem is PlacementUnitsCollapsiblePanel is reading player units from Swing thread without holding a game lock?

iirc my original implementation was careful to not have this problem. Looks like the change was:
https://github.com/triplea-game/triplea/commit/ad70991748ff0bddfa26d56e51fd1c7372ec275f

Before that, PlacePanel.updateStep() was calling UnitSeparator.categorize(playerUnits) block. So within that place, playerUnits can't change and categorize() returns a new object.

In the new code, in PlacementUnitsCollapsiblePanel - everything is being done in the invokeLater, which has problems like this - but also, the units may change before the swing code runs.

Dan, assigning to you since I believe your refactoring introduced this.

**Comment by DanVanAtta** [comment]
Please @ mention those that could have input on a bug. Once code is merged, it belongs to the project and all contributors to fix it. We must be striving for as bug-free merges as possible.

Assignment means active work is being done on a bug. Let's say someone makes 1000 changes, with 5% of them, it's not a good mechanism to then punish them to assign 50 items for them to fix alone. This process also breaks the paradigm of we treat everyone as if they are the casual committer. We have examples where people throw up PRs, but then are long gone and never contribute again. We cannot expect them to fix bugs, so our process needs to be flexible for this.

**Comment by DanVanAtta** [comment]
Hmm, that update is not a refactor. A refactor is a gang-of-four element and addresses code-smells. Generally it is a change that is done to code that has tests that allows those same tests to pass. Being pedantic there.

GameLocks are evil in TripleA.  When the game lock is released you can have updates at that time before updates are rendered. You need to be sure that gamelocks are always held long enough such that a result is both computed and rendered, not just during computation. Such locks introduce long delays and are likely unnecessary and buggy. 

In this case creating a copy of the data is probably a simple/good fix and will render accurate enough results. If the results are going to be re-rendered on update anyways, and that update is in the swing queue, blocking the swing queue long enough to then re-render immediately is seemingly pointless.

**Comment by DanVanAtta** [comment]
Sigh, code that is careful to depend on surrounding contexts and locks is brittle. That's in part why the locks are evil. The same code being used in two different places, or different times can then lead to breaks. That's just brittle.. 

Regardless, thanks for pinpointing the problem @asvitkine . I think a fix here is not too difficult, can PR it pretty soon.

**Comment by DanVanAtta** [comment]
A fix is up:  #7070
